net: lwm2m: Add Docker based test script

SeppoTakalo opened the pull request:
Add Docker based test script for LwM2M interoperability test.

Depends on https://github.com/zephyrproject-rtos/net-tools/pull/65

SeppoTakalo committed:
net: lwm2m: Add Docker based test script

Add Docker based test script for LwM2M interoperability test.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented on scripts/net/run-sample-tests.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
This was so minimal change that I did not create separate commit for it.

But if using minimal `/bin/sh` the script does not work. It seem to rely on some Bash specific `source SCRIPT` instead of `. SCRIPT`

SeppoTakalo commented:
I updated the manifest as well, as the net-tools change is now merged in.

SeppoTakalo commented:
Now runs fine against the `master` branch of net-tools.

```
❯ ./scripts/net/run-sample-tests.sh tests/net/lib/lwm2m/interop
$ZEPHYR_BASE /home/seta/src/ncs/zephyr
$NET_TOOLS_BASE /home/seta/src/ncs/tools/net-tools
Started Docker network 'net-tools0' bridge interface 'br-af606bb110c5'...
Started Docker container 'net-tools' with extra arguments '--ip=192.0.2.2 --ip6=2001:db8::2 -p 8080:8080 -p 8081:8081 -p 5683:5683/udp'...
Docker PID 822573
Running '/net-tools/start-leshan.sh' in the container...
INFO    - Using Ninja..
INFO    - Zephyr version: v3.5.0-rc1-99-gc0a5e8a763f2
DEBUG   - Running cmake script /home/seta/src/ncs/zephyr/cmake/verify-toolchain.cmake
DEBUG   - Calling cmake: /usr/bin/cmake -DFORMAT=json -P /home/seta/src/ncs/zephyr/cmake/verify-toolchain.cmake
DEBUG   - Finished running /home/seta/src/ncs/zephyr/cmake/verify-toolchain.cmake
INFO    - Using 'zephyr' toolchain.

... clip clip ...

DEBUG   - PYTEST: ------------------------------ live log teardown -------------------------------
DEBUG   - PYTEST: DEBUG: Running subprocess finished with return code 0
DEBUG   - PYTEST: - generated xml file: /home/seta/src/ncs/zephyr/tests/net/lib/lwm2m/interop/twister-out/native_posix/tests/net/lib/lwm2m/interop/net.lwm2m.interop/report.xml -
DEBUG   - PYTEST: ======================== 2 passed in 149.22s (0:02:29) =========================
DEBUG   - Could not find a matching testcase for net.lwm2m.interop.test_lwm2m_bootstrap_psk
DEBUG   - Could not find a matching testcase for net.lwm2m.interop.test_lwm2m_nosecure
DEBUG   - run status: native_posix/tests/net/lib/lwm2m/interop/net.lwm2m.interop passed
INFO    - 1/1 native_posix              tests/net/lib/lwm2m/interop/net.lwm2m.interop      PASSED (native 149.225s)

INFO    - 1 test scenarios (1 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
--------------------------------
Total test suites: 1
Total test cases: 2
Executed test cases: 2
Skipped test cases: 0
Completed test suites: 1
Passing test suites: 1
Failing test suites: 0
Skipped test suites: 0
Skipped test suites (runtime): 0
Skipped test suites (filter): 0
Errors: 0
--------------------------------
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 errored, 0 skipped with 0 warnings in 156.12 seconds
INFO    - In total 2 test cases were executed, 0 skipped on 1 out of total 626 platforms (0.16%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/seta/src/ncs/zephyr/tests/net/lib/lwm2m/interop/twister-out/twister.json
INFO    - Writing xunit report /home/seta/src/ncs/zephyr/tests/net/lib/lwm2m/interop/twister-out/twister.xml...
INFO    - Writing xunit report /home/seta/src/ncs/zephyr/tests/net/lib/lwm2m/interop/twister-out/twister_report.xml...
INFO    - Run completed
Stopping Docker PIDs 822573 
Sample 'interop' successful
Stopped Docker container 'net-tools'...
Stopped Docker network 'net-tools0' bridge interface 'br-af606bb110c5'...
```

I hope someone else verifies this as well so I don't accidentally rely on something that is installed on my machine, but not mentioned anywhere.
One missing package I noticed was a `bridge-utils` that was used in the net-tools when configuring docker network.

rlubos commented:
Are there any additional steps required besides running the script? I can't get a connection to the Leshan server when running locally:
```
DEBUG   - PYTEST: DEBUG: Starting new HTTP connection (1): localhost:8080
DEBUG   - PYTEST: SKIPPED (Leshan server not available)
```
I've tried running `mqtt_publisher` via docker and that one seems to work just fine.

Apart from that, two minor caveats:
* The net-tools discovery in https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/net/run-sample-tests.sh#L35 seems to work no more, as the net-tools moved to `tools/net-tools`. I've had to specify `NET_TOOLS_BASE` manually
* `lwm2m_observation.c` throws a warning, treated as an error (gcc version 13.1.1 20230429):
```
 337   │ /home/robert/repos/zephyrproject/zephyr/subsys/net/lib/lwm2m/lwm2m_observation.c: In function ‘en
       │ gine_observe_node_discover’:
 338   │ /home/robert/repos/zephyrproject/zephyr/subsys/net/lib/lwm2m/lwm2m_observation.c:546:28: error: ‘
       │ path_list_size’ may be used uninitialized [-Werror=maybe-uninitialized]
 339   │   546 |                         if (obs_list_size != path_list_size) {
 340   │       |                            ^
 341   │ /home/robert/repos/zephyrproject/zephyr/subsys/net/lib/lwm2m/lwm2m_observation.c:532:28: note: ‘p
       │ ath_list_size’ was declared here
 342   │   532 |         int obs_list_size, path_list_size;
 343   │       |
```

SeppoTakalo commented:
Did you update the `net-tools` image?

(Assuming you did the west update to get the master branch the west is pointing to)

```
❯ cd tools/net-tools/docker
❯ docker rmi net-tools
Untagged: net-tools:latest
Deleted: sha256:496eb7ce7aeee113eca7ac07f0a31941f5a00ce42dc97eddd06326ef168fa994
❯ docker build -t net-tools .
```

Then test:
```
cd zephyr
export NET_TOOLS_BASE=/home/seta/src/ncs/tools/net-tools
./scripts/net/run-sample-tests.sh tests/net/lib/lwm2m/interop
```

SeppoTakalo committed:
net: script: Fix the net-tools path

Fix the correct path to net-tools directory.
Also default to Bash instead of /bin/sh as we use
Bash syntax.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented:
I fixed the `net-tools` path in the script as well.

rlubos commented:
> Did you update the net-tools image?

Yes, just tried with a fresh docker image build to be 100% sure, but with the same otucome.

rlubos commented:
The issues I've been observing were related to some unreasonably high Leshan process start time in Docker, I suspect this must be something on my side, I don't think this should be a reason to block the PR. Running tests with docker manually works just fine.
